resource/resourceadapters/charmstore_test.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package resourceadapters_test
+
+import (
+	"fmt"
+
+	"github.com/juju/errors"
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/charmstore"
+	"github.com/juju/juju/resource/resourceadapters"
+)
+
+type CharmStoreSuite struct {
+	testing.IsolationSuite
+
+	resourceClient *testResourceClient
+}
+
+var _ = gc.Suite(&CharmStoreSuite{})
+
+func (s *CharmStoreSuite) SetUpTest(c *gc.C) {
+	s.IsolationSuite.SetUpTest(c)
+	s.resourceClient = &testResourceClient{
+		stub: &testing.Stub{},
+	}
+}
+
+func (s *CharmStoreSuite) TestGetResourceTerminates(c *gc.C) {
+	msg := "trust"
+	attempts := 0
+	s.resourceClient.getResourceF = func(req charmstore.ResourceRequest) (data charmstore.ResourceData, err error) {
+		attempts++
```

Comment:
will getResourceF be called via another goroutine? It might be worth atomic.AddInt(attempts, 1)
to ensure we don't have 'go test -race' issues.